Gems/EMotionFX/Code/Source/Editor/Plugins/Ragdoll/JointSwingLimitManipulators.cpp

```diff
@@ -0,0 +1,245 @@
+                BeginEditing();
+            });
+        m_swingZManipulator->InstallMouseMoveCallback(
+            [this](const AzToolsFramework::LinearManipulator::Action& action)
+            {
+                if (m_jointSwingLimitState.m_swingLimitZ.has_value())
+                {
+                    float newSwingLimitZ = ManipulatorScale * action.LocalPosition().GetY();
+                    m_physicsSetupManipulatorData.m_jointConfiguration->SetPropertyValue(AZ::Name("SwingLimitZ"), newSwingLimitZ);
+                    AZStd::optional<float> validatedSwingLimitZ =
+                        m_physicsSetupManipulatorData.m_jointConfiguration->GetPropertyValue(AZ::Name("SwingLimitZ"));
+                    if (validatedSwingLimitZ.has_value())
+                    {
+                        m_swingZManipulator->SetLocalPosition(
+                            AZ::Vector3(ManipulatorOffsetX, ManipulatorInverseScale * validatedSwingLimitZ.value(), 0.0f));
+                    }
+                    InvalidateEditorValues();
+                }
+            });
+        m_swingZManipulator->InstallLeftMouseUpCallback(
+            [this]([[maybe_unused]] const AzToolsFramework::LinearManipulator::Action& action)
+            {
+                FinishEditing();
+            });
+
+        Refresh();
+
+        AZ::TickBus::Handler::BusConnect();
+        PhysicsSetupManipulatorRequestBus::Handler::BusConnect();
+    }
+
+    void JointSwingLimitManipulators::Refresh()
+    {
+        if (m_physicsSetupManipulatorData.HasJointLimit())
+        {
+            const AZ::Transform parentWorldTransform = m_physicsSetupManipulatorData.m_nodeWorldTransform *
+                AZ::Transform::CreateFromQuaternion(m_physicsSetupManipulatorData.m_jointConfiguration->m_parentLocalRotation);
+
+            AZStd::optional<float> swingLimitY =
+                m_physicsSetupManipulatorData.m_jointConfiguration->GetPropertyValue(AZ::Name("SwingLimitY"));
+            if (swingLimitY.has_value())
+            {
+                m_swingYManipulator->SetSpace(parentWorldTransform);
+                m_swingYManipulator->SetLocalPosition(AZ::Vector3(ManipulatorOffsetX, 0.0f, ManipulatorInverseScale * swingLimitY.value()));
+            }
+            AZStd::optional<float> swingLimitZ =
+                m_physicsSetupManipulatorData.m_jointConfiguration->GetPropertyValue(AZ::Name("SwingLimitZ"));
+            if (swingLimitZ.has_value())
+            {
+                m_swingZManipulator->SetSpace(parentWorldTransform);
+                m_swingZManipulator->SetLocalPosition(AZ::Vector3(ManipulatorOffsetX, ManipulatorInverseScale * swingLimitZ.value(), 0.0f));
+            }
+        }
+    }
+
+    void JointSwingLimitManipulators::Teardown()
+    {
+        PhysicsSetupManipulatorRequestBus::Handler::BusDisconnect();
+        AZ::TickBus::Handler::BusDisconnect();
+        m_swingYManipulator->Unregister();
+        m_swingZManipulator->Unregister();
```

Comment:
question: Should we also call `.reset` on these too?

Reply:
oh yeah, good point.